File: src/types/api/benchmark.ts
import type { ModelType } from "./model";

export type BenchmarkStartInput = {
  provider_id: string;
  model_id?: string | null;
  dataset_id: string;
  mode: string;
  concurrency: number;
  duration_seconds: number;
  start_concurrency?: number;
  end_concurrency?: number;
  step_strategy?: "double" | "linear";
  step_value?: number;
  stage_sample_rounds?: number;
  stage_duration_seconds?: number;
  warmup_rounds?: number;
  warmup_seconds?: number;
  request_timeout_seconds?: number;
  sla_p95_ms?: number;
  min_success_rate?: number;
  sla_stop_policy?: "continue_full_staircase" | "stop_on_failure";
  workload_config?: BenchmarkWorkloadConfig;
  request_log_config?: RequestLogConfig;
};

export type RequestLogConfig = {
  enabled: boolean;
  capture_body: boolean;
  max_records_per_stage: number;
};

export type BenchmarkWorkloadConfig = {
  streaming?: boolean;
  max_output_tokens?: number;
  prompt_profile?: "short" | "mixed" | "long";
  batch_size?: number;
  text_count_per_request?: number;
  documents_per_query?: number;
  top_k?: number;
  image_profile?: "small" | "medium" | "large";
  image_count?: number;
};

export type BenchmarkTaskSummary = {
  id: string;
  name: string;
  status: string;
  model_type: ModelType | string;
  model_name: string;
  provider_name: string;
  dataset_name: string;
  concurrency: number;
  success_rate: number;
  p95_latency_ms: number;
  goodput_qps: number;
  created_at: string;
};

export type StopResult = {
  task_id: string;
  stopped: boolean;
};

export type MetricsTick = {
  task_id: string;
  elapsed_seconds: number;
  qps: number;
  latency_ms: number;
  ttft_ms: number;
  tps: number;
  success_rate: number;
  errors: number;
  in_flight: number;
  request_count: number;
  success_count: number;
  failure_count: number;
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  batch_size: number;
  text_count: number;
  documents_per_query: number;
  pair_count: number;
  image_count: number;
};

export type StageChangedEvent = {
  task_id: string;
  stage: string;
  message: string;
  stage_index?: number | null;
  stage_total?: number | null;
  concurrency?: number | null;
};

export type BenchmarkRequestLogPageInput = {
  task_id: string;
  page: number;
  page_size: number;
  stage_index?: number;
  status?: string;
  keyword?: string;
};

export type BenchmarkRequestLogSummary = {
  id: string;
  task_id: string;
  stage_index: number;
  request_index: number;
  sample_index: number;
  status: string;
  latency_ms: number;
  ttft_ms: number;
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  error_kind?: string | null;
  prompt_preview?: string | null;
  response_preview?: string | null;
  created_at: string;
};

export type BenchmarkRequestLogDetail = BenchmarkRequestLogSummary & {
  prompt?: string | null;
  response_text?: string | null;
  raw_error?: string | null;
  raw_usage?: unknown;
  body_available: boolean;
};

export type BenchmarkRequestLogPage = {
  items: BenchmarkRequestLogSummary[];
  total: number;
  page: number;
  page_size: number;
};

File: src-tauri/src/models/benchmark.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct BenchmarkStartInput {
    pub provider_id: String,
    pub model_id: Option<String>,
    pub dataset_id: String,
    pub mode: String,
    pub concurrency: i64,
    pub duration_seconds: i64,
    pub start_concurrency: Option<i64>,
    pub end_concurrency: Option<i64>,
    pub step_strategy: Option<String>,
    pub step_value: Option<i64>,
    pub stage_sample_rounds: Option<i64>,
    pub stage_duration_seconds: Option<i64>,
    pub warmup_rounds: Option<i64>,
    pub warmup_seconds: Option<i64>,
    pub request_timeout_seconds: Option<i64>,
    pub sla_p95_ms: Option<i64>,
    pub min_success_rate: Option<f64>,
    pub sla_stop_policy: Option<String>,
    pub workload_config: Option<serde_json::Value>,
    pub request_log_config: Option<RequestLogConfig>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct RequestLogConfig {
    pub enabled: bool,
    pub capture_body: bool,
    pub max_records_per_stage: i64,
}

impl RequestLogConfig {
    pub fn normalized(input: Option<&Self>) -> Self {
        let Some(input) = input else {
            return Self {
                enabled: false,
                capture_body: false,
                max_records_per_stage: 200,
            };
        };
        Self {
            enabled: input.enabled,
            capture_body: input.enabled && input.capture_body,
            max_records_per_stage: input.max_records_per_stage.clamp(1, 1000),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct BenchmarkTaskSummary {
    pub id: String,
    pub name: String,
    pub status: String,
    pub model_type: String,
    pub model_name: String,
    pub provider_name: String,
    pub dataset_name: String,
    pub concurrency: i64,
    pub success_rate: f64,
    pub p95_latency_ms: i64,
    pub goodput_qps: f64,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct StopResult {
    pub task_id: String,
    pub stopped: bool,
}

#[derive(Debug, Clone, Serialize)]
pub struct MetricsTick {
    pub task_id: String,
    pub elapsed_seconds: i64,
    pub qps: f64,
    pub latency_ms: i64,
    pub ttft_ms: i64,
    pub tps: f64,
    pub success_rate: f64,
    pub errors: i64,
    pub in_flight: i64,
    pub request_count: i64,
    pub success_count: i64,
    pub failure_count: i64,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub total_tokens: i64,
    pub batch_size: i64,
    pub text_count: i64,
    pub documents_per_query: i64,
    pub pair_count: i64,
    pub image_count: i64,
}

#[derive(Debug, Clone)]
pub struct BenchmarkErrorRecord {
    pub task_id: String,
    pub error_kind: String,
    pub message: String,
    pub count: i64,
}

#[derive(Debug, Clone, Serialize)]
pub struct BenchmarkRequestLogSummary {
    pub id: String,
    pub task_id: String,
    pub stage_index: i64,
    pub request_index: i64,
    pub sample_index: i64,
    pub status: String,
    pub latency_ms: i64,
    pub ttft_ms: i64,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub total_tokens: i64,
    pub error_kind: Option<String>,
    pub prompt_preview: Option<String>,
    pub response_preview: Option<String>,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct BenchmarkRequestLogDetail {
    #[serde(flatten)]
    pub summary: BenchmarkRequestLogSummary,
    pub prompt: Option<String>,
    pub response_text: Option<String>,
    pub raw_error: Option<String>,
    pub raw_usage: Option<serde_json::Value>,
    pub body_available: bool,
}

#[derive(Debug, Clone, Deserialize)]
pub struct BenchmarkRequestLogPageInput {
    pub task_id: String,
    pub page: i64,
    pub page_size: i64,
    pub stage_index: Option<i64>,
    pub status: Option<String>,
    pub keyword: Option<String>,
}

impl BenchmarkRequestLogPageInput {
    pub fn normalized(&self) -> Self {
        let page_size = match self.page_size {
            20 | 50 | 100 | 200 => self.page_size,
            value if value <= 0 => 50,
            value => value.min(200),
        };
        Self {
            task_id: self.task_id.clone(),
            page: self.page.max(1),
            page_size,
            stage_index: self.stage_index,
            status: self
                .status
                .as_ref()
                .map(|value| value.trim().to_string())
                .filter(|value| !value.is_empty()),
            keyword: self
                .keyword
                .as_ref()
                .map(|value| value.trim().to_string())
                .filter(|value| !value.is_empty()),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct BenchmarkRequestLogPage {
    pub items: Vec<BenchmarkRequestLogSummary>,
    pub total: i64,
    pub page: i64,
    pub page_size: i64,
}

#[derive(Debug, Clone)]
pub struct BenchmarkRequestLogRecord {
    pub summary: BenchmarkRequestLogSummary,
    pub body_ref: Option<String>,
    pub prompt: Option<String>,
    pub response_text: Option<String>,
    pub raw_error: Option<String>,
    pub raw_usage: Option<serde_json::Value>,
}
